### DISCUSSION TITLE
An easy-to-use Dec App framework design

### DISCUSSION THREAD

**Discussion by ghost** [buckyos/CYFS#248] [Ideas]
After I learned the current design concept of CYFS Dec App, I have a stage summary of it:
1. a Dec App consists of two components, Service and Web
2. the Dec App Service runs on ood all the time and requires a process-level single instance system
3. Dec App Service registers the post object handler with Gateway Service and processes the object passed by the handler through a callback.
4. Dec App Web runs in the CYFS browser and sends requests to Dec App Service through the post_object interface of CYFS Runtime. The request is encapsulated in an Object

Here I have an idea:
Is it possible to make the development of DecApp easier with a DecApp middle layer, which we can temporarily call "DecNginx"?

The basic idea is as follows:
- Simplify the communication between Dec App Web and Dec App Service to a url+body form similar to HTTP Get/Post.
- Describe which requests DecApp has to handle through a configuration file like nginx.conf
- There is a middleware?wrapper?sdk?-like component that simplifies post_object
- on the web side, use interfaces like dec_nginx.get(), dec_nginx.post() to access the dec service
- On the dec service side, use an interface like dec_nginx.on(callback) to handle requests in a similar way to an http server
- Or the dec service side can even support a per request per process approach, where the middle tier calls up the configured script when the request arrives and processes the request

**Comment by lurenpluto** [comment]
In fact, this idea similar to Nginx was considered a long time ago, but now, to add a similar "DecNginx", we might as well consider running a complete nginx service directly between cyfs-stack and dec, so that the dec service developed with a traditional web server can run directly behind nginx. The dec service can run directly behind nginx

Because nginx is mature and easy to use, if we support nginx, then the development of many dec services will become very simple, and we don't have to repeat the wheel and design a similar "DecNginx", which causes additional learning and maintenance costs.

 
![DecNginx](https://user-images.githubusercontent.com/3624316/236411394-f668ef71-830c-4b52-8dfe-199048d77443.png)
